js/style/style_batch.js

```diff
@@ -46,11 +48,22 @@ styleBatch.prototype = {
         if (this._style._layers[layer.id] !== undefined) {
             throw new Error('There is already a layer with this ID');
         }
+
+
         if (!(layer instanceof StyleLayer)) {
+            if (validateStyle.emitErrors(this._style, validateStyle.layer({
+                value: layer,
+                style: this._style.serialize(),
+                styleSpec: styleSpec
+            }))) return this;
```

Comment:
It's tricky, because this batch should work:

1. Start condition: no source "foo"
2. Add source "foo"
3. Add layer with source "foo"

But this batch should fail:

1. Start condition: has source "bar"
2. Remove source "bar"
3. Add layer with source "bar"